src/Common/src/TypeSystem/Common/TypeSystemHelpers.cs

```diff
@@ -36,18 +36,24 @@ static public PointerType MakePointerType(this TypeDesc type)
 
         static public DefType GetClosestDefType(this TypeDesc type)
         {
-            if (type is DefType)
-                return (DefType)type;
-            else
-                return type.BaseType;
+            // The implementation for this should mirror GetClosestMetadataType
+            throw new NotImplementedException();
         }
 
         static public MetadataType GetClosestMetadataType(this TypeDesc type)
         {
-            if (type is MetadataType)
-                return (MetadataType)type;
-            else
-                return type.BaseType.GetClosestMetadataType();
+            if (type.IsSzArray)
+            {
+                MetadataType arrayType = type.Context.SystemModule.GetType("System", "Array`1");
```

Comment:
I don't want ArrayType.GetMethods to return any of the explicitly implemented interface methods. They are a hidden implementation detail of ArrayType, and in most ways the method's aren't really part of arrays... GetClosestMetadataType is bit of a weird beast. I'm not at this point familiar with what uses we have for it. Certainly our virtual function resolution logic depends on it, but I'm not sure if anything else does. I like your idea of having this function call a virtual on TypeSystemContext to get at the Array<T> type, but it doesn't really provide as much abstraction as one would like. (For instance, it cannot represent the CLR handling of this where all of the interesting instance methods end up implemented as generic instance methods of a non generic type.